write/writed.8
.\" Copyright (c) 1980, 1991 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" Portions of the code documented by this man page are 
.\" Copyright (C) 1991 by the Massachusetts Institute of Technology.
.\" Please see the file COPYING for details.
.\"
.\"
.\"     from: @(#)fingerd.8	6.4 (Berkeley) 3/16/91
.\"     from: fingerd.8,v 1.1 1994/06/08 15:12:24 rzsfl Exp rzsfl
.\"	$Id: writed.8,v 1.1 1996/11/23 19:50:40 dholland Exp $
.\"
.\" Rewritten as writed.8, Apr 29 1996, dholland@hcs.harvard.edu
.\"
.Dd April 29, 1996
.Dt WRITED 8
.Os "Linux 1.3"
.Sh NAME
.Nm writed
.Nd remote write server
.Sh SYNOPSIS
.Nm writed
.Sh DESCRIPTION
.Nm Writed
implements a very simple protocol that allows a 
.Xr write 1
program on one host to connect up to one on another host, allowing
communications across the network. The details of this protocol
are intentionally not described in this man page in order to 
discourage fraudulent usage. This constitutes a bug. 
.Pp
.Nm Writed
is meant to be run from 
.Xr inetd 8
listening for
.Tn TCP
requests on port 811.
.Sh SEE ALSO
.Xr write 1
.Sh BUGS
Connecting directly to the server from a
.Tn TIP
or an equally narrow-minded
.Tn TELNET Ns \-protocol
user program can result
in meaningless attempts at option negotiation being sent to the
server, which will foul up the command line interpretation.
.Nm Writed
should be taught to filter out
.Tn IAC Ns \'s
and perhaps even respond
negatively
.Pq Tn IAC WON'T
to all option commands received.
.Pp
No attempt is made to determine the authenticity of users sending messages.
.Sh HISTORY
The
.Nm
command was developed at MIT in conjunction with Project Athena and
is essentially a hackup of
.Xr fingerd 8 .


netwrite/writed/writed.c
/*
 * This code is copyrighted. Please see the file COPYING for a full notice.
 */

/*
 * writed - Receive network write connections with inetd.
 *
 * Note:  This is mostly fingerd.c with all occurances of
 * "finger" changed to "write". Well, not so much any more.
 *
 * $Log: writed.c,v $
 * Revision 1.7  1999/10/02 01:46:45  netbug
 * some extranious header files commented out
 * added socklen_t to MCONFIG.in and to writed.c
 *
 * Revision 1.6  1999/08/01 00:05:24  dholland
 * Use new version of confgen. Set version to 0.15. Update README.
 *
 * Revision 1.5  1997/06/09 01:32:03  dholland
 * minor glibc fixes
 *
 * Revision 1.4  1997/03/08 17:24:30  dholland
 * Oops, fix warning.
 *
 * Revision 1.3  1997/03/08 12:42:04  dholland
 * Don't depend on gcc extensions
 *
 * Revision 1.2  1996/11/25 18:43:05  dholland
 * clean compile.
 *
 * Revision 1.1  1996/11/23  19:50:40  dholland
 * Initial revision
 *
 * Revision 1.8  1996/04/29  21:43:25  dholland
 * Copyright fixes.
 *
 * Revision 1.7  1996/04/29  20:25:43  dholland
 * Config improvements.
 *
 * Revision 1.6  1996/04/29  19:59:14  dholland
 * Add RCS stuff.
 *
 */

char copyright[] =
  "@(#) Copyright (c) 1983 Regents of the University of California.\n"
  "All rights reserved.\n";
/*
 * From: /afs/rel-eng.athena.mit.edu/project/release/current/
 *        source/bsd-4.3/common/etc/RCS/writed.c,v 
 *   1.3 90/04/05 18:31:44 epeisach Exp
 */
char rcsid[] = "$Id: writed.c,v 1.7 1999/10/02 01:46:45 netbug Exp $";
#include "../version.h"

/*
 * Write server.
 */
/* #include <mit-copyright.h> */

#include <sys/types.h>
#include <netinet/in.h>
#include <netdb.h>

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>

#ifdef ultrix  /* missing protos */
struct sockaddr;
int getpeername(int fd, struct sockaddr *, int *len);
#endif

/* Should be set by makefile */
#ifndef BINDIR
#define BINDIR "/usr/local/bin"
#endif

#define WS " \t\r\n"
#define AC 6  /* max # of args to write, incl terminating null */

/*
#define DEBUG
*/

static void fatal2(char *t) {
  printf("%s\n", t);
  exit(1);
}

static void fatal(char *s, char *t) {
  if (errno!=0) printf("%s: %s\n", s, strerror(errno));
  fatal2(t);
}

static void secure(const char *fromarg) {
  struct hostent *ho;
  struct sockaddr_in sin;
  socklen_t x = sizeof(sin);
  char *u, *h, *t;
  u = strdup(fromarg);
  if (!u) fatal("strdup", "Out of memory");
  h = strchr(u, '@');
  if (!h) fatal2("Usage: write user [tty]");
  *h++=0;
  t = strchr(h, '@');
  if (t) {
    *t++=0;
    if (strchr(t, '@')) fatal2("Usage: write user [tty]");
  }

  ho = gethostbyname(h);
  if (!ho) fatal("gethostbyname", "Where are you?");
  if (getpeername(0, (struct sockaddr *) &sin, &x) < 0) {
    fatal("getpeername", "Where are you?");
  }
  free(u);

  for (x=0; ho->h_addr_list[x]; x++) {
#ifdef DEBUG
    unsigned char *p = ho->h_addr_list[x];
    unsigned char *q = (char *) &sin.sin_addr;
    printf("Testing: %u.%u.%u.%u vs %u.%u.%u.%u [%d]\n", 
	   p[0], p[1], p[2], p[3], q[0], q[1], q[2], q[3], ho->h_length);
    fflush(stdout);
#endif
    if (ho->h_length > (int)sizeof(sin.sin_addr)) {
	ho->h_length = sizeof(sin.sin_addr);
    }
    if (!memcmp(ho->h_addr_list[x], (char *) &sin.sin_addr, ho->h_length))
      return;
  }
  fatal2("Host name lookup error");
}


int main(void /*int argc, char *argv[]*/) {
  char line[BUFSIZ], *av[AC];
  int i=2;
  av[0] = "write";
  av[1] = "-f";
  *line=0;
#ifdef DEBUG
  printf("writed ready\n");
  fflush(stdout);
#endif
  fgets(line, BUFSIZ, stdin);
  for (av[i]=strtok(line, WS); av[i++] && i<AC; av[i]=strtok(NULL, WS));
  av[AC-1] = NULL;
  dup2(0, 1);
  dup2(0, 2);
  secure(av[2]);
#ifdef DEBUG
  printf("Ok, all clear\n");
  fflush(stdout);
#endif
  execv(BINDIR "/write", av);
  execv("/usr/local/bin/write", av);
  execv("/usr/bin/write", av);
  execv("/bin/write", av);
  fatal("execv", "Can't find binary for write");
  return 0;
}
